appinterface/projection/crossfire/crossfire.go

```diff
@@ -0,0 +1,754 @@
+			}
+
+			if err := crossfireValidatorsView.Upsert(&validatorRow); err != nil {
+				return fmt.Errorf("error inserting new validator into view: %v", err)
+			}
+
+			// checkTaskSetup
+			if err := projection.checkTaskSetup(
+				crossfireValidatorsView,
+				validatorRow.OperatorAddress,
+				validatorRow.ConsensusNodeAddress,
+				validatorRow.JoinedAtBlockTime,
+			); err != nil {
+				return fmt.Errorf("error check Setup task for new validator: %v", err)
+			}
+		}
+	}
+	for _, event := range events {
+		if msgSubmitSoftwareUpgradeProposalEvent, ok := event.(*event_usecase.MsgSubmitSoftwareUpgradeProposal); ok {
+			projection.logger.Debug("handling MsgSubmitSoftwareUpgradeProposal event")
+
+			// Check if proposed after competition has ended
+			if blockTime.After(projection.competitionEndTime) {
+				return fmt.Errorf("error Competition has already ended")
+			}
+
+			// Check if proposed before OR after Phase 2
+			if blockTime.Before(projection.phaseTwoStartTime) || blockTime.After(projection.phaseThreeStartTime) {
+				return fmt.Errorf("error This proposal does not occur in Phase 2")
+			}
+
+			// Check if proposed by NOT an admin
+			if msgSubmitSoftwareUpgradeProposalEvent.ProposerAddress != projection.adminAddress {
+				return fmt.Errorf("error checking proposer address equals admin address")
+			}
+
+			// Check if proposal ID does not match the required ID
+			if msgSubmitSoftwareUpgradeProposalEvent.MaybeProposalId != nil && *msgSubmitSoftwareUpgradeProposalEvent.MaybeProposalId != projection.networkUpgradeProposalID {
+				return fmt.Errorf("error checking Proposal ID in proposal")
+			}
+
+			networkUpgradeTimestamp := msgSubmitSoftwareUpgradeProposalEvent.MsgSubmitSoftwareUpgradeProposalParams.Content.Plan.Time
+			networkUpgradeBlockheight := msgSubmitSoftwareUpgradeProposalEvent.MsgSubmitSoftwareUpgradeProposalParams.Content.Plan.Height
+
+			// Update Network upgrade target Timestamp in DB
+			networkUpgradeTimestampDBKey := constants.NETWORK_UPGRADE_TARGET_TIMESTAMP_KEY()
+			errTSUpdate := crossfireChainStatsView.Set(networkUpgradeTimestampDBKey, networkUpgradeTimestamp.UnixNano())
+			if errTSUpdate != nil {
+				return fmt.Errorf("error updating network_upgrade timestamp: %v", errTSUpdate)
+			}
+
+			// Update Network upgrade target Blockheight in DB
+			networkUpgradeBlockheightDBKey := constants.NETWORK_UPGRADE_TARGET_BLOCKHEIGHT_KEY()
+			errBlockheightUpdate := crossfireChainStatsView.Set(networkUpgradeBlockheightDBKey, networkUpgradeBlockheight)
+
+			if errBlockheightUpdate != nil {
+				return fmt.Errorf("error updating network_upgrade blockheight: %v", errBlockheightUpdate)
+			}
+
+		} else if msgVoteCreatedEvent, ok := event.(*event_usecase.MsgVote); ok {
+			projection.logger.Debug("handling MsgVote event")
+
+			// Check if proposed after competition has ended
+			if blockTime.After(projection.competitionEndTime) {
+				return fmt.Errorf("error Competition has already ended")
```

Comment:
It is possible that the proposal is proposed by someone else, and the voter votes for it. Here we should just ignore it.